codec-http/src/main/java/io/netty/handler/codec/http/HttpServerUpgradeHandler.java

```diff
@@ -313,32 +313,32 @@ private boolean upgrade(final ChannelHandlerContext ctx, final FullHttpRequest r
 
         // Create the user event to be fired once the upgrade completes.
         final UpgradeEvent event = new UpgradeEvent(upgradeProtocol, request);
+        final ChannelFuture writeComplete = ctx.writeAndFlush(upgradeResponse);
+
+        // After writing the upgrade response we immediately prepare the
+        // pipeline for the next protocol to avoid a race between completion
+        // of the write future and receiving data before the pipeline is
+        // restructured.
+        try {
+            // Perform the upgrade to the new protocol.
+            sourceCodec.upgradeFrom(ctx);
+            upgradeCodec.upgradeTo(ctx, request);
+
+            // Remove this handler from the pipeline.
+            ctx.pipeline().remove(HttpServerUpgradeHandler.this);
+
+            // Notify that the upgrade has occurred. Retain the event to offset
+            // the release() in the finally block.
+            ctx.fireUserEventTriggered(event.retain());
```

Comment:
Is there a reason to delay this until the write succeeds? (I can't think of any)

Reply:
The write hasn't necessarily succeeded yet as `writeComplete` is a `ChannelPromise`. We should enqueue the write before reshaping the pipeline as the incoming codec _may_ try to write something when it gets added to the pipeline.

Reply:
Maybe just delay the write by using ctx.eventLoop().execute(...) and schedule the write via it ?

Reply:
That did the trick, thanks @normanmaurer. I'd tried to defer the satisfaction of the promise via indirection through the executor, but the `flush()` call would run it. Your suggestion is much more reasonable anyway.

Reply:
Ah, shoot, my comment was really unclear - I wanted to note that we `fireUserEventTriggered` before the write is complete and ask if there could be a reason to delay that event until the write finished instead. I'm not sure what the contract around that user event is, but maybe some folks rely on that happening after the write?